### PR TITLE
Feat: Applied Stops to controllers

### PR DESCRIPTION
**Pull request by aagusuab**
### Description of your changes

Fixes #

I have:

- [x] Read and followed Caravel's [Code of conduct](https://github.com/Azure/k8s-work-api/blob/master/code-of-conduct.md).
- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

**Review by ryanzhang-oss — COMMENTED** [review]
Please do not work on this

File: pkg/controllers/finalize_controller_unit_test.go [hunk 1]
```diff
@@ -49,6 +52,32 @@ func TestGarbageCollectAppliedWork(t *testing.T) {
 	}
 }
 
+func TestFinalizerReconcile(t *testing.T) {
+	tests := map[string]struct {
+		r              FinalizeWorkReconciler
+		expectedResult ctrl.Result
+		expectedError  error
+	}{
+		"Happy Path: AppliedWork deleted, work finalizer removed": {
```
Comment (Arvindthiru) [inline pkg/controllers/finalize_controller_unit_test.go:61]: Change test case name, you are testing the reconcile not working

File: pkg/controllers/finalize_controller_unit_test.go [hunk 2]
```diff
@@ -49,6 +52,32 @@ func TestGarbageCollectAppliedWork(t *testing.T) {
...
+			r:              FinalizeWorkReconciler{Joined: false},
+			expectedResult: ctrl.Result{RequeueAfter: time.Minute * 5},
+			expectedError:  fmt.Errorf("finalize controller is not started yet"),
+		},
+	}
+	for testName, tt := range tests {
+		t.Run(testName, func(t *testing.T) {
+			ctrlResult, err := tt.r.Reconcile(ctx, ctrl.Request{
```
Comment (Arvindthiru) [inline pkg/controllers/finalize_controller_unit_test.go:69]: Add the second variable as a field in the tests hashmap above

File: pkg/controllers/work_status_controller_unit_test.go [hunk 3]
```diff
@@ -1,13 +1,16 @@
 package controllers
 
 import (
-	"testing"
-
+	"context"
```
Comment (Arvindthiru) [inline pkg/controllers/work_status_controller_unit_test.go:4]: Fix imports

File: pkg/controllers/finalize_controller_unit_test.go [hunk 4]
```diff
@@ -49,6 +52,32 @@ func TestGarbageCollectAppliedWork(t *testing.T) {
 	}
 }
 
+func TestFinalizerReconcile(t *testing.T) {
```
Comment (Arvindthiru) [inline pkg/controllers/finalize_controller_unit_test.go:55]: Rename to TestFinalizerReconcileStop ?

File: pkg/controllers/finalize_controller_unit_test.go (hunk 1 above)
Comment (aagusuab) [reply pkg/controllers/finalize_controller_unit_test.go]: Changed wrong test case names to correct ones.

File: pkg/controllers/finalize_controller_unit_test.go (hunk 2 above)
Comment (aagusuab) [reply pkg/controllers/finalize_controller_unit_test.go]: I have added ctrl.Request as a variable, and used it. 
However,  I did not include it in the hashmap, as if I were to have more reconciliation tests, they would use same requests.

File: pkg/controllers/work_status_controller_unit_test.go (hunk 3 above)
Comment (aagusuab) [reply pkg/controllers/work_status_controller_unit_test.go:4]: Fixed

File: pkg/controllers/finalize_controller_unit_test.go (hunk 4 above)
Comment (aagusuab) [reply pkg/controllers/finalize_controller_unit_test.go:55]: I will leave this as a Reconcile, as in the future, we could have more tests related to Reconcile function in the finalizer.

File: pkg/controllers/finalize_controller_unit_test.go (hunk 2 above)
Comment (Arvindthiru) [reply pkg/controllers/finalize_controller_unit_test.go]: The general rule I heard from Ryan is to have all the input variables in the struct, I understand the global variable part

File: pkg/controllers/apply_controller.go [hunk 5]
```diff
@@ -84,6 +86,11 @@ type applyResult struct {
 func (r *ApplyWorkReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	klog.InfoS("Work apply controller reconcile loop triggered.", "item", req.NamespacedName)
 
+	if !r.Joined {
+		klog.InfoS("work controller is not started yet")
+		return ctrl.Result{RequeueAfter: time.Minute * 5}, fmt.Errorf("work controller is not started yet")
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/controllers/apply_controller.go:91]: yeah, more like 5 seconds

[the remark this answers is not in the document]
Comment (aagusuab) [reply pkg/controllers/apply_controller.go]: changed to 5 seconds.

**Review by Arvindthiru — APPROVED** [review]

File: pkg/controllers/apply_controller.go [hunk 6]
```diff
@@ -84,6 +86,11 @@ type applyResult struct {
 func (r *ApplyWorkReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	klog.InfoS("Work apply controller reconcile loop triggered.", "item", req.NamespacedName)
 
+	if !r.Joined {
+		klog.InfoS("work controller is not started yet")
+		return ctrl.Result{RequeueAfter: time.Second * 5}, fmt.Errorf("work controller is not started yet")
```
Comment (ryanzhang-oss) [inline pkg/controllers/apply_controller.go:91]: error takes precedence over requeue

**Pull request merged**